### ISSUE TITLE
min and max

### ISSUE THREAD

**Issue by Pjb518** [UnicornHeartClub/Issues#64]
Currently min and max have a bizarre implementation which is really unintuitive. At present, they basically redefine the range on a die, so `1d8max16` effectively become `1d16`. A proposed use for this is FATE dice, which require a -1 or a 1. Implementation for that is currently `1d3min-1max1rr0`, which limits the range from -1 to 1 (-1, 0, 1), and then forces rerolls of zero until a nonzero value is rolled.

I propose that we replace the current min and max implementation with a special 1dx roll, which can take custom values as an array. FATE dice would then be implemented using `1dx[-1, 1]`, removing the need for the rerolling 0, and hugely simplifying the result. This 1dx system also allows for a much greater range of options, including unfair dice, dice missing specific values, or dice that have nonnumerical values. In fact, I see no real reason to limit the content of the array at all, and it could easily take variables or other rolls.

min and max could then perform a far more intuitive task, specifically outputting a minimum or maximum value when a roll exceeds the defined range. For example, I might want to set a roll to ignore any rolls of a 1, replacing the roll with a 2 instead. Some 5e feats give a bonus of this sort. This does **not** permit a reroll, and instead returns a consistent value. I might write this example as `1d10min2`, which would return a minimum value of 2 instead of changing the number of sides on the die. There are still 10 results on this die, but, essentially, two of them are 2s now.

**Comment by Aquaricat** [comment]
Agreed - implementations of min/max in dice rollers is pretty straight forward.  Using the 1dx format to demonstrate
* `1d8min5` becomes `1dx[5,5,5,5,5,6,7,8]`
* `1d8max5` becomes `1dx[1,2,3,4,5,5,5,5]`

For example, I'd use the min roll for rolling cleric HP where the DM wants to make the minimum HP gain equal to the default of 5.

The current implementation - where `1d8max9` becomes `1dx[1,2,3,4,5,6,7,8,9]` is very confusing.  Firstly, I can't imagine a situation where I wouldn't just enter `1d9`.  It's easier to read, and easier to understand.  Secondly, it breaks the expectation that the overall odds of rolling a given number on a die roll are the same, but that a *floor* or *ceiling* in the result has been added to the results.

**Comment by Pjb518** [comment]
To make it easier to set up custom dice using the proposed 1dx, I think there should be some shorthand added down the line for fairly common series, which will generate the array for you. This might look something like this:
```bash
!r 1dx[<initial value> -> <end value> step:<step value>]

!r 1dx[1 -> 9 step: 2] = an array of 1, 3, 5, 7, 9
```
We could also support some reserved words like step: even (include all even numbers in the range), step: odd (same for odd numbers).

**Comment by Pjb518** [comment]
New 1d[x, y, z] syntax implemented and working for most cases. Keeping this open due to the unusual min and max flags which are now redundant and should be repurposed as per the suggestions above.
